tests/integration/dump/test_dump_dbms.py

```diff
@@ -183,6 +198,76 @@ def error_callback(resp):
         assert_dump_data(dump_path, docs, shards, pea_id)
 
 
+@pytest.mark.parametrize('nr_docs', [700])
+@pytest.mark.parametrize('emb_size', [10])
+def test_threading_query_while_reloading(tmpdir, nr_docs, emb_size):
+    docs = list(get_documents(nr=nr_docs, index_start=0, emb_size=emb_size))
+    assert len(docs) == nr_docs
+    nr_search = 3
+
+    def _validate_results_empty(resp):
+        assert len(resp.docs[0].matches) == 0
+
+    def _validate_results_nonempty(resp):
+        assert len(resp.docs) == nr_search
+        for d in resp.docs:
+            if nr_docs < 10:
+                # using np.testing since pytest+pycharm swallow the stack info on built-in assert
+                np.testing.assert_equal(len(d.matches), nr_docs)
+            else:
+                # TODO does it return all of them no matter how many?
+                np.testing.assert_(len(d.matches) > 0, 'no matches')
+            for m in d.matches:
+                np.testing.assert_equal(m.embedding.shape[0], emb_size)
+                assert (
+                    DBMSIndexDriver._doc_without_embedding(m).SerializeToString()
+                    is not None
+                )
+                assert 'hello world' in m.text
+                assert f'tag data' in m.tags['tag_field']
+
+    def error_callback(resp):
+        raise Exception('error callback called')
+
+    dump_path = os.path.join(str(tmpdir), 'dump_dir')
+    os.environ['DBMS_WORKSPACE'] = os.path.join(str(tmpdir), 'index_ws')
+    os.environ['QUERY_WORKSPACE'] = os.path.join(str(tmpdir), 'query_ws')
+
+    os.environ['USES_AFTER'] = '_pass'
+    os.environ['QUERY_SHARDS'] = str(1)
+
+    with Flow.load_config('flow_dbms.yml') as flow_dbms:
+        with Flow.load_config('flow_query.yml') as flow_query:
+            with TimeContext(f'### indexing {len(docs)} docs'):
+                flow_dbms.index(docs)
+
+            with TimeContext(f'### dumping {len(docs)} docs'):
+                flow_dbms.dump('indexer_dbms', dump_path, shards=1, timeout=-1)
+
+            dir_size = path_size(dump_path)
+            print(f'### dump path size: {dir_size} MBs')
+
+            # test with query while reloading async.
+            t = Thread(
+                target=flow_query.rolling_update, args=('indexer_query', dump_path)
+            )
+            # searching on the still empty replica
+            t.start()
+            flow_query.search(
+                docs[:nr_search],
+                on_done=_validate_results_empty,
+                on_error=error_callback,
+            )
+            t.join()
+
+            # done with both -- we should have matches now
+            flow_query.search(
+                docs[:nr_search],
+                on_done=_validate_results_nonempty,
```

Comment:
use mockers to ensure callbacks are indeed called

Reply:
done